parser.py

```diff
@@ -1,17 +1,49 @@
 #!/bin/env python
 import argparse
 import pprint
+import re
+from abc import abstractmethod
+from typing import override
+
 import grammar
 import visitor
 
 
+class Preprocessor:
+    @abstractmethod
+    def process(self, program: str):
+        pass
+
+
+class CommentsPreprocessor(Preprocessor):
+
+    @override
+    def process(self, program: str):
+        cleaned_lines = filter(
+            lambda cleaned_line: cleaned_line,
+            map(lambda line: re.sub(r'#.*$', '', line), program.splitlines()))
+        return "\n".join(cleaned_lines)
+
+
+class PreprocessorPipeline:
```

Comment:
This is a nice idea. I would go even further and make it implement the `Preprocessor` interface as well. Also these three classes could probably live in a `preprocessor.py` file.

Reply:
I thought about the interface, but I didn't really like the fact that the pipeline is not, as such, an implementation of the preprocessor. I don't know, I don't really have a final preference on this subject.

Reply:
It is indeed an implementation of the preprocessor as the preprocessor is an interface that gets some text an proprocesses it and that's exactly what the pipeline does. It simply uses a special strategy to accomplish it based on calling other Preprocessor implementors. :vulcan_salute: